src/Take.Blip.Builder/StateManager.cs

```diff
@@ -12,44 +14,89 @@ public class StateManager : IStateManager
         private const string PREVIOUS_STATE_PREFIX = "previous";
         private const string STATE_ID_KEY = "stateId";
 
-        public Task<string> GetStateIdAsync(IContext context, CancellationToken cancellationToken)
+        private const int STATE_ID_INDEX = 0;
+        private const int PREVIOUS_STATE_ID_INDEX = 1;
+        private const string STATE_AND_PREVIOUS_SEPARATOR = ";";
+
+        private readonly IConfiguration _configuration;
+
+        public StateManager(
+           IConfiguration configuration
+           )
         {
-            return context.GetContextVariableAsync(GetStateKey(context.Flow.Id), cancellationToken);
+            _configuration = configuration;
         }
 
-        public Task<string> GetPreviousStateIdAsync(IContext context, CancellationToken cancellationToken)
+        public async Task<string> GetStateIdAsync(IContext context, CancellationToken cancellationToken) => ExtractStateId(await GetStateIdContextVariable(context, cancellationToken));
+
+        public async Task<string> GetPreviousStateIdAsync(IContext context, CancellationToken cancellationToken)
         {
-            return context.GetContextVariableAsync(GetPreviousStateKey(context.Flow.Id), cancellationToken);
+            var previousStateId = ExtractPreviousStateId(await GetStateIdContextVariable(context, cancellationToken));
+
+            //using GetPreviousStateKey to compatibility with previous versions
+            return previousStateId.IsNullOrEmpty() ? await context.GetContextVariableAsync(GetPreviousStateKey(context.Flow.Id), cancellationToken) : previousStateId;
         }
 
-        public Task<string> GetParentStateIdAsync(IContext context, CancellationToken cancellationToken)
+        public async Task<string> GetParentStateIdAsync(IContext context, CancellationToken cancellationToken)
         {
             if (context.Flow.Parent == null)
             {
                 return null;
             }
 
-            return context.GetContextVariableAsync(GetStateKey(context.Flow.Parent.Id), cancellationToken);
+            return ExtractStateId(await GetParentStateIdContextVariable(context, cancellationToken));
         }
 
-        public Task SetStateIdAsync(IContext context, string stateId, CancellationToken cancellationToken)
+        public Task SetStateIdAsync(IContext context, string stateId, string previousStateId, CancellationToken cancellationToken)
```

Comment:
Use `async/await` instead of returning `Tasks`'s directly.
See https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-asyncawait-over-directly-returning-task